polkadot/node/network/collator-protocol/src/validator_side/claim_queue_state/basic.rs

```diff
@@ -0,0 +1,2231 @@
+	/// are removed.
+	///
+	/// Note 2: During pruning we remove all the candidates for the pruned relay parent because we
+	/// no longer need to know about them. If the claim was not undone so far - it will be
+	/// permanent.
+	candidates_per_rp: HashMap<Hash, HashSet<CandidateHash>>,
+}
+
+impl ClaimQueueState {
+	/// Create an empty `ClaimQueueState`. Use [`add_leaf`] to populate it.
+	pub(crate) fn new() -> Self {
+		Self {
+			block_state: VecDeque::new(),
+			future_blocks: VecDeque::new(),
+			candidates_per_rp: HashMap::new(),
+		}
+	}
+
+	pub(super) fn fork(&self, target_relay_parent: &Hash) -> Option<Self> {
+		if self.block_state.back().and_then(|state| state.hash) == Some(*target_relay_parent) {
+			// don't fork from the last block!
+			return None
+		}
+
+		// Find the index of the target relay parent in the old state
+		let target_index = self
+			.block_state
+			.iter()
+			.position(|claim_info| claim_info.hash == Some(*target_relay_parent))?;
+
+		let block_state =
+			self.block_state.iter().cloned().take(target_index + 1).collect::<VecDeque<_>>();
+
+		let rp_in_view = block_state.iter().filter_map(|c| c.hash).collect::<HashSet<_>>();
+
+		let candidates = self
+			.candidates_per_rp
+			.iter()
+			.filter(|(rp, _)| rp_in_view.contains(rp))
+			.map(|(rp, c)| (*rp, c.clone()))
+			.collect::<HashMap<_, _>>();
+
+		let candidates_in_view = candidates.iter().fold(HashSet::new(), |mut acc, (_, c)| {
+			acc.extend(c.iter().cloned());
+			acc
+		});
+
+		// Transfer any claims from the target relay parent's ancestors onto the new path.
+		// Example:
+		// old_state: [A B C D]; target_relay_parent = B;
+		// Any claims on C and D should also be transferred to the new fork. Because when we claim a
+		// slot at relay parent X we claim it at all possible forks.
+		// Also note that only claims for candidates which fall within new fork's view are
+		// transferred.
+		let future_blocks = self
+			.get_window(target_relay_parent)
+			.skip(1)
+			.map(|c| ClaimInfo {
+				hash: None,
+				claim: c.claim,
+				claim_queue_len: 1,
+				claimed: c.claimed.clone_or_default(&candidates_in_view),
+			})
+			.collect::<VecDeque<_>>();
+
+		Some(ClaimQueueState { block_state, future_blocks, candidates_per_rp: candidates })
+	}
+
+	/// Appends a new leaf with its corresponding claim queue to the state.
+	pub(crate) fn add_leaf(&mut self, hash: &Hash, claim_queue: &VecDeque<ParaId>) {
+		if self.block_state.iter().any(|s| s.hash == Some(*hash)) {
+			return
+		}
+
+		// First check if our view for the future blocks is consistent with the one in the claim
+		// queue of the new block. If not - the claim queue has changed for some reason and we need
+		// to readjust our view.
+		for (idx, expected_claim) in claim_queue.iter().enumerate() {
+			match self.future_blocks.get_mut(idx) {
+				Some(future_block) =>
+					if future_block.claim.as_ref() != Some(expected_claim) {
```

Comment:
This is really not supposed to happen at all. I would suggest adding a debug_assert and at least a debug log if it happens - warning is actually better.